cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -224,6 +224,41 @@ Status ResolveOneFieldRef(
   return Status::OK();
 }
 
+bool IsNamedFieldRef(const FieldRef& ref) {
+  if (ref.IsName()) return true;
+  if (const auto* nested_refs = ref.nested_refs()) {
+    for (const auto& nested_ref : *nested_refs) {
+      if (!nested_ref.IsName()) return false;
+    }
+    return true;
+  }
+  return false;
+}
```

Comment:
Moved it into `FieldRef` in the update. Looking at it now though, I suspect it may be too niche to justify its place there - at least on its own.

Methods that transform a ref into a flat vector of names or indices might be more useful in general (but less trivial, of course).

Reply:
I don't think it's too niche.  I feel like I have run into situations a few times now in the scanner where I've needed to know if a ref is all-names, all-indices, or mixed (a lot of the new scanner stuff normalizes to all-indices).  We do have FieldPath already which is a flat vector of indices.